salt/grains/core.py

```diff
@@ -984,6 +984,60 @@ def _virtual(osdata):
     return grains
 
 
+def _virtual_hv(osdata):
+    '''
+    Returns detailed hypervisor information from sysfs
+    Currently this seems to be used only by Xen
+    '''
+    grains = {}
+
+    # Bail early if we're not running on Xen
+    if 'xen' not in osdata['virtual']:
+        return grains
+
+    # Try to get the exact hypervisor version from sysfs
+    try:
+        version = {}
+        for fn in ('major', 'minor', 'extra'):
+            with salt.utils.files.fopen('/sys/hypervisor/version/{}'.format(fn), 'r') as fhr:
+                version[fn] = fhr.read().strip()
+        grains['virtual_hv_version'] = '{}.{}{}'.format(version['major'], version['minor'], version['extra'])
+        grains['virtual_hv_version_info'] = (version['major'], version['minor'], version['extra'])
```

Comment:
We should be using a list instead of a tuple here, for a couple reasons:

1) The tuple will be lost (converted to list) anyway if/when the data is serialized using msgpack, json, etc.
2) We allow grains to be added via YAML, which does not support tuples.